### DISCUSSION TITLE
setValue doesnt work in useEffect

### DISCUSSION THREAD

**Discussion by deadly8** [react-hook-form/react-hook-form#4122] [Q&A]
Hello, I need to update my form value on initial render with the data that I receive from api.
I know about this issue https://github.com/react-hook-form/react-hook-form/issues/2578 and usage of 
`shouldUnregister: false`
but it causes me bug while revalidating fields that I create from` useFieldArray`

Is there any other way, how can I set values using setValue in useEffect hook?

I'm using v6.11.3

**Comment by wdfinch** [comment]
I would recommend using `reset` if you need to bulk update form values.

**Comment by deadly8** [comment]
@wdfinch do you mean use `reset` in useEffect before setValue? Can you show some example?

**Comment by wdfinch** [comment]
https://codesandbox.io/s/react-hook-form-controller-template-forked-mh9wf?file=/src/index.js

**Comment by deadly8** [comment]
yes result is good but validation is broken. I have `isValid` from `formState` always as false but data in form is present

**Comment by deadly8** [comment]
it's like RHF doesn't detect changes and didn't run validation itself
https://codesandbox.io/s/react-hook-form-controller-template-forked-3bqg1

**Comment by wdfinch** [comment]
Validation is fine. You have the validation mode set to `onChange`. Reset does **not** trigger validation. It's not an input `onChange` event which is your criteria.

If you want to trigger validation manually after you `setValues`/`reset` you can run `trigger()`. Note that reset and setValue are async so you must run trigger on the next render cycle after the internal `formState` has been updated. 

https://codesandbox.io/s/react-hook-form-controller-template-forked-pmre3?file=/src/index.js

**Comment by bluebill1049** [comment]
Thanks @wdfinch's help!
